File: services/src/main/java/com/latinos/services/remote/manager/NetworkManagerImpl.java
package com.latinos.services.remote.manager;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import com.latinos.services.remote.manager.error.RestNetworkException;

public class NetworkManagerImpl implements NetworkManager {

    /**
     * Checks if there is any active network
     */
    private static boolean isInternetAvailable(Context context) {
        NetworkInfo activeNetwork = getActiveNetwork(context);
        return activeNetwork != null && activeNetwork.isConnectedOrConnecting();
    }

    /**
     * Checks if the active connection is {@link ConnectivityManager#TYPE_WIFI}
     */
    private static boolean isWifiConnected(Context context) {
        NetworkInfo activeNetwork = getActiveNetwork(context);
        return activeNetwork != null && activeNetwork.getType() == ConnectivityManager.TYPE_WIFI;
    }

    /**
     * Get the active {@link NetworkInfo}
     */
    @SuppressLint("MissingPermission")
    private static NetworkInfo getActiveNetwork(Context context) {
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        return cm != null ? cm.getActiveNetworkInfo() : null;
    }

    private Context context;

    public NetworkManagerImpl(Context context) {
        this.context = context;
    }

    /**
     * Checks if Internet is available or throws a {@link RestNetworkException}
     *
     * @throws RestNetworkException when internet is not available
     */
    public void checkConnectivity() throws RestNetworkException {
        if (!isInternetAvailable(context)) {
            throw new RestNetworkException("Not Available");
        }
    }

    /**
     * Returns if the active network is {@link ConnectivityManager#TYPE_WIFI}. If there is no internet will throw a {@link RestNetworkException}
     *
     * @throws RestNetworkException when internet is not available
     */
    public boolean checkWifi() throws RestNetworkException {
        checkConnectivity();
        return isWifiConnected(context);
    }
}

File: services/src/main/java/com/latinos/services/remote/manager/NetworkManagerImpl.kt
package com.latinos.services.remote.manager

import android.annotation.SuppressLint
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import com.latinos.services.remote.manager.error.RestNetworkException

class NetworkManagerImpl(private val context: Context) : NetworkManager {
    /**
     * Checks if Internet is available or throws a [RestNetworkException]
     *
     * @throws RestNetworkException when internet is not available
     */
    @Throws(RestNetworkException::class)
    override fun checkConnectivity() {
        if (!isInternetAvailable(context)) {
            throw RestNetworkException("Not Available")
        }
    }

    /**
     * Returns if the active network is [ConnectivityManager.TYPE_WIFI]. If there is no internet will throw a [RestNetworkException]
     *
     * @throws RestNetworkException when internet is not available
     */
    @Throws(RestNetworkException::class)
    override fun checkWifi(): Boolean {
        checkConnectivity()
        return isWifiConnected(context)
    }

    companion object {
        /**
         * Checks if there is any active network
         */
        private fun isInternetAvailable(context: Context): Boolean {
            val activeNetwork = getActiveNetwork(context)
            return activeNetwork != null && activeNetwork.isConnectedOrConnecting
        }

        /**
         * Checks if the active connection is [ConnectivityManager.TYPE_WIFI]
         */
        private fun isWifiConnected(context: Context): Boolean {
            val activeNetwork = getActiveNetwork(context)
            return activeNetwork != null && activeNetwork.type == ConnectivityManager.TYPE_WIFI
        }

        /**
         * Get the active [NetworkInfo]
         */
        @SuppressLint("MissingPermission")
        private fun getActiveNetwork(context: Context): NetworkInfo? {
            val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            return cm?.activeNetworkInfo
        }
    }
}